### ISSUE TITLE
[Bug] firecracker crashes in new pid namespace

### ISSUE THREAD

**Issue by KarthikNedunchezhiyan** [firecracker-microvm/firecracker#2344]
## Describe the bug

Firecracker crashed if it is spawned after unshare `CLONE_NEWPID`. Crashed with below error.

```
2020-12-12T11:35:45.259888476 [anonymous-instance:ERROR:src/firecracker/src/main.rs:277] Building VMM configured from cmdline json failed: Internal(VcpuHandle(VcpuSpawn(Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" })))
```

## To Reproduce
`unshare -p ./firecracker --config-file fc_config.json --no-api`

content of fc_config.json:

```
{
    "boot-source": {
        "kernel_image_path": "hello-vmlinux.bin",
        "boot_args": "console=ttyS0 reboot=k panic=1 pci=off"
    },
    "drives": [
        {
            "drive_id": "rootfs",
            "path_on_host": "hello-rootfs.ext4",
            "is_root_device": true,
            "is_read_only": false
        }
    ]
}
```

## Expected behavior

Expected to run fine in new PID namespace.

## Environment

`[Author TODO: Please supply the following information):]`
`[ - Firecracker version.]`
Firecracker v0.23.1

`[ - Host kernel versions.]`
4.4.0-184-generic

`[ - Guest kernel and Rootfs used.]`
kernel and rootfs provided in the getting-started doc

`[ - Architecture.]`
Intel

## Additional context

`[Author TODO: What are you trying to achieve?]`
Trying to perform as instructed in jailer documentation regarding PID namespace.

## Checks
- [x] Have you searched the Firecracker Issues database for similar problems?
- [x] Have you read the existing relevant Firecracker documentation?
- [ ] Are you certain the bug being reported is a Firecracker issue?

**Comment by KarthikNedunchezhiyan** [comment]
Failed in API server mode too.

```
$ unshare -p ./firecracker --api-sock /tmp/firecracker.socket

2020-12-18T08:11:39.948598495 [anonymous-instance:ERROR:src/firecracker/src/main.rs:52] Firecracker panicked at 'API thread spawn failed.: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }', src/firecracker/src/api_server_adapter.rs:149:5
Aborted (core dumped)
```

I can notice some issues around `clone` syscall while spawning in PID namespace. In config-file mode issue occurred while spawning vcpu thread (clone syscall involved) and In API-server mode issue occurred while spawning api-server thread (clone syscall involved).

```
clone(child_stack=0x7f9f9eeab998, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID|0x400000, parent_tidptr=0x7f9f9eeabb58, tls=0x7f9f9eeabb20, child_tidptr=0x7c08e0) = -1 EINVAL (Invalid argument)
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
munmap(0x7f9f9eca9000, 2109440)         = 0
close(4)                                = 0
open("/etc/localtime", O_RDONLY|O_NONBLOCK|O_CLOEXEC) = -1 ENOENT (No such file or directory)
write(2, "2020-12-18T08:14:30.296414607 [a"..., 2682020-12-18T08:14:30.296414607 [anonymous-instance:ERROR:src/firecracker/src/main.rs:52] Firecracker panicked at 'API thread spawn failed.: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }', src/firecracker/src/api_server_adapter.rs:149:5
) = 268
```

**Comment by KarthikNedunchezhiyan** [comment]
Some interesting findings

```
fn main() {
    std::thread::spawn(||{
        println!("hello from new thread");
    }).join().unwrap();
}
```

above code failed with the same error when using `unshare -p ./binary`

**Comment by acatangiu** [comment]
This is normal. `Firecracker` starts multiple threads:
 - emulation thread (main thread)
 - api thread
 - multiple vcpu threads (one for each configured vcpu)

`unshare CLONE_NEWPID` disallows the process to spawn threads so firecracker will obviously fail

**Comment by KarthikNedunchezhiyan** [comment]
As like @acatangiu pointed, after `unshare CLONE_NEWPID` we are not allowed to spawn a thread. Below is the issue that I made on unshare, how it failed and how to fix that. This info may help other readers.

Issue made:
`unshare` command tool directly exec to the given firecracker binary after `unshare CLONE_NEWPID` syscall. At this point we still not made `fork` or `clone` so, the firecracker crashed while spawning a thread. [pid_namespace [under Compatibility of CLONE_NEWPID with other CLONE_* flags]](https://man7.org/linux/man-pages/man7/pid_namespaces.7.html) has explained why spawning a thread before `clone` or `fork` is restricted.

Fix:
unshare -fp ./firecracker --config-file fc_config.json --no-api

I have added `f` option additionally. Now unshare command tool will fork and spawn the firecracker binary instead of exec.
